src/csharp/Microsoft.Spark.Worker/Command/SqlCommandExecutor.cs

```diff
@@ -153,22 +162,22 @@ internal class PicklingSqlCommandExecutor : SqlCommandExecutor
         /// </summary>
         /// <param name="stream">Stream to write to</param>
         /// <param name="rows">Rows to write to</param>
-        private void WriteOutput(Stream stream, IEnumerable<object> rows)
+        /// <param name="estimatedMaxSize">Estimated max size of the serialized output</param>
+        private void WriteOutput(Stream stream, IEnumerable<object> rows, int estimatedMaxSize)
         {
-            MemoryStream writeOutputStream = s_writeOutputStream ??
-                (s_writeOutputStream = new MemoryStream());
-            writeOutputStream.Position = 0;
+            if (s_output == null)
+                s_output = new byte[estimatedMaxSize];
 
             Pickler pickler = s_pickler ?? (s_pickler = new Pickler(false));
-            pickler.dump(rows, writeOutputStream);
+            pickler.dumps(rows, ref s_output, out int bytesWritten);
 
-            if (writeOutputStream.Position == 0)
+            if (bytesWritten == 0)
             {
                 throw new Exception("Message buffer cannot be null.");
```

Comment:
Not really your change - but this exception message isn't illustrative of what happened. Would it be possible to fix this message while you are here?

Reply:
Good point, I have updated the message.